site/source/tutorials/005_form_widgets/index.md

```diff
@@ -58,107 +77,128 @@ Let's use those classes and a few virtual DOM nodes, to add the visual elements
 			v('fieldset', { classes: this.classes(css.nameField) }, [
 				v('legend', { classes: this.classes(css.nameLabel) }, [ 'Name' ]),
 				w(TextInput, {
-					key: 'input1',
+					key: 'firstNameInput',
 					label: {
 						content: 'First Name',
 						hidden: true
 					},
 					placeholder: 'Given name',
-					type: 'text' as 'text'
+					required: true
 				}),
 				w(TextInput, {
-					key: 'input2',
+					key: 'lastNameInput',
 					label: {
 						content: 'Last Name',
 						hidden: true
 					},
-					placeholder: 'Family name',
-					type: 'text' as 'text'
+					placeholder: 'Surname name',
+					required: true
 				})
 			]),
 			w(TextInput, {
-				key: 'input3',
 				label: 'Email address',
-				type: 'email'
+				type: 'email',
+				required: true
 			}),
-			w(Button, {
-				content: 'Save',
-				type: 'submit'
-			})
+			w(Button, { content: 'Save' })
 		]);
 	}
 ```
 
-At this point, the user interface for the form is available, but it does not do anything since we have not specified any event handlers. In the [last tutorial](../004_user_interactions/), we learned how to add event handlers to custom widgets by assigning a method to an event. When using pre-built widgets, however, we pass the handlers as properties. For example, we are going to need a way to handle each text field's `change` event. To do that, we provide the desired handler function as the `onChange` property that is passed to the widget. Update the `render` method once again:
+At this point, the user interface for the form is available, but it does not do anything since we have not specified any event handlers. In the [last tutorial](../004_user_interactions/), we learned how to add event handlers to custom widgets by assigning a method to an event. When using pre-built widgets, however, we pass the handlers as properties. For example, we are going to need a way to handle each text field's `input` event. To do that, we provide the desired handler function as the `onInput` property that is passed to the widget.
 
-{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:33-110 %}
+{% instruction 'Update the `render` method once again.' %}
 
-This form of the `render` method now does everything that we need from: it creates the user interface and registers the event handlers that will update the application as the user enters information.
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:43-86 %}
 
-The `render` method starts by decomposing the properties into local constants. We still need to define those properties, so update the `WorkerFormProperties` interface to include them:
+This form of the `render` method now does everything that we need from: it creates the user interface and registers the event handlers that will update the application as the user enters information. However, we need to add a few more methods to the `WorkerForm` to define the event handlers.
```

Comment:
"This form of the `render` method now does everything that we need from: " ...missing a word, and could be more terse perhaps